MAINT: Refactor updateifcopy

mattip opened the pull request:
Replaces pull requests #9269 and #9451, see extensive previous discussions there. Summary:
- moves the code to resolve ``UPDATEIFCOPY`` semantics from ``array_dealloc`` into a new API function, call the new function both from ``array_dealloc`` and from all the places that previously relied on ``Py_DECREF``. This is to support implementations with a different GC strategy, where ``Py_DECREF`` may not immediately cause calling ``array_dealloc`` and ``UPDATEIFCOPY`` resolution
- In order to alert downstream NumPy C-API consumers who use the ``UPDATEIFCOPY`` semantics of this change, add a new flag ``NPY_ARRAY_WRITEBACKIFCOPY`` and a new function ``PyArray_SetWritebackIfCopyBase``. Use of either of those will raise a ``FutureWarning``.
- This pull request does not touch nditer, with its own special opflag ``updateifcopy``.That will be handled in a future pull request. As  a consequence a warning on resolution of the ``UPDATEIFCOPY`` semantics in ``array_dealloc`` will only happen now if the compiler flag ``-DDEPRECATE_UPDATEIFCOPY`` is used. On PyPy that is the default.

mattip committed:
ENH: add API to resolve UPDATEONCOPY outside dealloc, test and use

charris commented:
You should close older PRs if you intend to replace them with this. You can always resubmit them.

mattip committed:
UPDATEIFCOPY -> WRITEBACKIFCOPY, documentation

mattip commented:
@njsmith could you review? I reworked the pull request:
- there is now ``WRITEBACKIFCOPY`` with the single letter ``X`` in flags, use of the old ``UPDATEIFCOPY (U)`` will issue a ``FutureWarning``. 
- Added tests for the new flag and that the old one does indeed issue a warning, 
- Addeda test for a ``FUTUREWARNING`` if using ``UPDATEIFCOPY`` in the array creation routines.
- Document the new behavior and deprecation of ``UPDATEIFCOPY``

njsmith commented on doc/CAPI.rst.txt:
```diff
@@ -274,19 +274,23 @@ array, or part of some larger record array.  But, they may have other uses...
     this flag to be set in such a situation). The base attribute then
     points to the "misbehaved" array (which is set read_only). When
-    the array with this flag set is deallocated, it will copy its
+    ``PyArray_ResoveWritebackIfCopy`` is called, it will copy its
```
missing `l` in `Resolve`.

Text should probably say something like "if you use this flag, you have to later call `PyArray_ResolveWriteBackIfCopy` on the array. If the the array has the `NPY_WRITEBACKIFCOPY` flag set, it will copy its contents back ... Otherwise, this operation is a no-op."

njsmith commented on doc/CAPI.rst.txt:
```diff
@@ -274,19 +274,23 @@ array, or part of some larger record array.  But, they may have other uses...
+``NPY_UPDATEIFCOPY``
+    Similar to ``NPY_WRITEBACKIFCOPY``, but deprecated since it copied the
+    contents back when the array is deallocated, which is not explicit and
+    relies on refcount semantics.
```
Should probably go ahead and mention PyPy here by name, since not everyone knows why refcount semantics matter :-). "which is unreliable on alternative Python implementations like PyPy that don't use refcounting.", maybe.

njsmith commented on numpy/add_newdocs.py:
```diff
@@ -4289,16 +4295,17 @@ def luf(lamdaexpr, *args, **kwargs):
-    ultimate owner of the memory exposes a writeable buffer interface,
-    or is a string. (The exception for string is made so that unpickling
-    can be done without copying memory.)
+    The (deprecated) UPDATEIFCOPY and WRITEBACKIFCOPY flags can never be set
```
"The WRITEBACKIFCOPY and (deprecated) UPDATEIFCOPY", to make clear that we're not saying both are deprecated.

njsmith commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -894,11 +895,11 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
 #define NPY_ARRAY_OUT_ARRAY    (NPY_ARRAY_CARRAY)
 #define NPY_ARRAY_INOUT_ARRAY  (NPY_ARRAY_CARRAY | \
-                                NPY_ARRAY_UPDATEIFCOPY)
+                                NPY_ARRAY_WRITEBACKIFCOPY)
```
Unfortunately it turns out third party code is using `INOUT_ARRAY` (e.g. `scipy/interpolate/src/_interpolate.cpp`), so I guess we can't just change its value :-(. We could define some new version like `NPY_ARRAY_INOUT_ARRAY2`, or maybe we should just deprecate it and tell people that they should write `NPY_ARRAY_CARRY | NPY_ARRAY_WRITEBACKIFCOPY` if that's what they want.

mattip replied:
For backward compatibility I prefer the first option. I will extend the warning notice to reflect the possible flags that could trigger it.

njsmith commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -894,11 +895,11 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
 #define NPY_ARRAY_OUT_FARRAY   (NPY_ARRAY_FARRAY)
 #define NPY_ARRAY_INOUT_FARRAY (NPY_ARRAY_FARRAY | \
-                                NPY_ARRAY_UPDATEIFCOPY)
+                                NPY_ARRAY_WRITEBACKIFCOPY)
```
Whatever we do for `NPY_ARRAY_INOUT_ARRAY` we should do the same here.

mattip committed:
fixes for review

mattip commented:
changes pushed in response to review, once approved I will squash to a single commit

njsmith commented on numpy/core/src/multiarray/cblasfuncs.c:
```diff
@@ -457,7 +457,7 @@ cblas_matrixproduct(int typenum, PyArrayObject *ap1, PyArrayObject *ap2,
             Py_DECREF(ap2);
             Py_DECREF(out_buf);
-            return PyArray_Return(result);
+            return PyArray_Return(out_buf);
```
What's this change about? It looks pretty dubious...

(This looks like it might be related to a recent fix to `dot` on master -- maybe the `+` line here is what master used to have, and somehow in your rebasing you ended up reverting it?)

mattip replied:
Good catch

njsmith commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -208,18 +208,35 @@ arrayflags_dealloc(PyArrayFlagsObject *self)
+arrayflags_updateifcopy_get(PyArrayFlagsObject *self)
+{
+    PyObject *item;
+    if(DEPRECATE_FUTUREWARNING("UPDATEIFCOPY deprecated, use WRITEBACKIFCOPY instead") < 0) {
```
Why is this a `FutureWarning` instead of a deprecation?

mattip committed:
review2, fix new test

njsmith commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -2923,7 +2923,7 @@ npyiter_allocate_arrays(NpyIter *iter,
             if (op_itflags[iop] & NPY_OP_ITFLAG_WRITE) {
                 Py_INCREF(op[iop]);
-                if (PyArray_SetUpdateIfCopyBase(temp, op[iop]) < 0) {
+                if (PyArray_SetWritebackIfCopyBase(temp, op[iop]) < 0) {
```
Is this change and the next one, inside the `nditer` code, correct? I thought we were keeping `nditer` using `UPDATEIFCOPY` for now?

mattip replied:
My goal was for ``nditer`` to not print any warning after this pull request on CPython. The following logic allows that:
- all calls to ``PyArray_SetUpdateIfCopyBase`` issue a ``FutureWarning``,
- to maintain backward compatiblity, ``PyArray_ResolveWritebackIfCopy`` is called in ``array_dealloc`` if ``arr->base`` exists
  - if ``PyArray_ResolveWritebackIfCopy``actually does something, it will currently warn, but only on PyPy or if compiled with ``-DDEPRECATE_UPDATEIFCOPY" 

By using ``PyArray_SetWritebackIfCopyBase``, ``nditer`` will neither warn when creating its operands nor when resolving in ``array_dealloc``. When we deal with ``nditer`` in a future pull request, we can enable the warning in ``array_dealloc`` on all implementations.

njsmith replied:
> all calls to `PyArray_SetUpdateIfCopyBase` issue a `FutureWarning`

This should be a `DeprecationWarning`. (Rule: `DeprecationWarning` is for when we're trying to remove an API, `FutureWarning` is for when we're going to keep it but change the semantics.) And I think this is where we should do the "if PyPy or `-DDEPRECATE_UPDATEIFCOPY`" check.

> `array_dealloc`

I think I'd prefer the `array_dealloc` logic be:

* If `WRITEBACKIFCOPY` is set, then issue a `RuntimeWarning` and that's it (you're supposed to call `ResolveUpdateIfCopy` or `ClearUpdateIfCopy` before the array is GCed).

* If `UPDATEIFCOPY` is set, then resolve it up like we always did. We don't need to issue a warning, because we already did that when `UPDATEIFCOPY` was set.

This then suggests that `nditer` should keep using `UPDATEIFCOPY` until we're ready to fix it properly, which seems logical, since it will in fact be broken until then, and if you're on PyPy or have `-DDEPRECATE_UPDATEIFCOPY` set then it's probably useful to see that your program is in fact using `UPDATEIFCOPY` (via nditer).

njsmith commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -89,14 +89,23 @@ def test_writeable(self):
+            assert_equal(self.a.flags['U'], False)
+            assert len(sup) == 2
+        assert_equal(self.a.flags.writebackifcopy, False)
+        assert_equal(self.a.flags['X'], False)
```
Can you use `numpy.testing.assert_warns` here?

mattip replied:
then I should wrap each test in a ``assert_warns`` context manager? (likewise line 4323)

njsmith replied:
Something like that? Honestly I don't remember the details, I just remember that @seberg had to go on like some ridiculous epic quest in order to get reliable warning tests working, so we should make use of his work :-). (@seberg, any hints?)

seberg replied:
Nothing special there, the epic quest is just to never use "ignore" filter (and we have a test to make sure you don't do that). Assert warns as context filter is great, if you want to use the message text, you can use `suppress_warnings` directly.

njsmith commented:
OK, I finally reached the bottom! Sorry for the drawn-out review :-)

Overall it's looking quite reasonable. What state do you think it's at -- ready to go in?

The one remaining high level thing that occurred to me is that `PyArray_XDECREF_ERR` is pretty weird with the new semantics. Maybe we should have a `PyArray_ClearWritebackIfCopyBase` function clears the flag and `->base`, basically the inverse of `SetWritebackIfCopyBase`?

mattip commented:
@njsmith thanks for the non-trivial review and discussion process. Except for the ``assert_warns`` tweak, it seems ready, and does what I expect on PyPy as well.  

If I understand correctly, ``PyArray_XDECREF_ERR`` seems to be a last-ditch effort to salvage ``arr->base`` if something else crashed, and ``arr`` should be discarded. IMO it should not touch ``arr`` at all, since its state is screwed up anyway, and it is about to be thrown away, so the less code the better.

What next? Should I squash the commits to a single one?

njsmith commented:
Re: `PyArray_XDECREF_ERR`: my point is that with the old `UPDATEIFCOPY` API, the way you eventually resolve the whole view/copy/UPDATEIFCOPY situation is to either call `Py_DECREF` or  `PyArray_XDECREF_ERR` on the array. This is problematic and makes the API design part of my brain itch, because it overloads the normal DECREF semantics to have this other meaning, but at least it's consistent – you always DECREF and the DECREF is always overloaded to resolve the `UPDATEIFCOPY`.

With this PR currently, the `WRITEBACKIFCOPY` API is that you have to eventually call either `PyArray_ResolveWritebackIfCopy` or else `PyArray_XDECREF_ERR`. This is weird and inconsistent – what does DECREF have to do with anything? It could work, but my preference would be to clean it up by deprecating `PyArray_XDECREF_ERR` and replacing it with a `PyArray_DiscardWritebackIfCopy` or similar that's like `PyArray_ResolveWritebackIfCopy` in that neither of them do the DECREF.

mattip committed:
fix new test for using self.assert_deprecated

mattip committed:
change deprecation logic as per review

mattip committed:
new logic exposed places where PyArray_ResolveWritebackIfCopy not called

mattip commented:
Changing the logic exposed a missing call to ``PyArray_ResolveWritebackIfCopy`` in the ufunc iterator code. Now fixed.

The ``PyArray_XDECREF_ERR`` ->  ``PyArray_DiscardWritebackIfCopy`` change is still not in the pull request

MSeifert04 committed:
BUG: Fix usage of keyword "from" as argument name for "can_cast".

Also removed inconsistency between the second argument name
between documentation ("totype") and code ("to").

mattip committed:
deprecate PyArray_XDECREF_ERR in favor of PyArray_DiscardWritebackIfCopy

mattip commented:
@njsmith rebased on master (it seems I am off-by-one) and added a commit to deprecate ``PyArray_XDECREF_ERR``. What is the next stage?

mattip commented:
ping, can we move forward with this?

mattip commented:
bitrot is beginning to set in. The current conflict is simple to solve, but what is considered best practice for updating pull requests? If I rebase, all the previous code comments will get lost

mattip commented:
@njsmith ping

mattip commented:
Can I at least get a 1.14 tag for this so it doesn't totally get lost?
Anyone?

mattip commented:
is there someone else who could pick up the review of this branch and help me get it polished enough to merge?

xoviat commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -85,17 +85,56 @@ PyArray_Size(PyObject *op)
+    #define DEPRECATE_UPDATEIFCOPY
+  #endif
+#endif
+
```
This seems odd to me. Why are we only deprecating this for PyPy? And IIUC, `DEPRECATE` should always be called.

mattip replied:
See the design [comment](https://github.com/numpy/numpy/pull/9639#issue-254637365)  at the top of the pull request, and issue #9714 which covers the problem in depth. We cannot yet issue this warning because it will be emitted by ``nditer( ..., [updateifcopy], ...)``, with no option to silence it. Once this pull request lands, I will turn my attention to solving issue #9714, hopefully by choosing one of the alternatives mentioned there

xoviat replied:
Either the warning can be issued or it cannot. If it cannot, then surely pypy will have the same problem?

xoviat replied:
Ordinarily, I would say that you should rename the function so that internal users can use it without emitting the warning but external users cannot. But that may increase merge conflicts, and I know what it's like to be worried about having to rebase.

njsmith replied:
The PR also moves all the internal users it can over to the new API, so there's no need for a no-warning version of the old API.

mattip replied:
Until issue #9714 is resolved, ``npyiter_allocate_arrays()`` in ``nditer_constr.c`` still uses the old interface (i.e. calls this function), so yes, there is still a need for a no-warning version of the old API on CPython

It is hard enough to make this enormous change in NumPy without adding in yet another dimension of pain in fixing #9714, which will require [engaging with the community](http://numpy-discussion.10968.n7.nabble.com/nditer-and-updateifcopy-semantics-advice-needed-td44739.html) to try to determine if and how to fix it 

However, IMO in the mean time, users of PyPy should be warned that using the semantics in #9714 on PyPy is not well supported, and could cause wrong answers. It seems this compromise is the least evil of all the alternatives.

xoviat replied:
> It is hard enough to make this enormous change in NumPy without adding in yet another dimension of pain in fixing #9714, which will require engaging with the community to try to determine if and how to fix it

+1. If this requires additional refactoring to avoid the warning, I don't think you should do it for now.

> However, IMO in the mean time, users of PyPy should be warned that using the semantics in #9714 on PyPy is not well supported, and could cause wrong answers. It seems this compromise is the least evil of all the alternatives.

I disagree. I think we should leave the warning as a comment for now because there are still internal users. I don't think numpy should try to behave differently for a different implementation.

njsmith replied:
@xoviat: also logic on warning on PyPy first is that right now, the remaining `nditer` case gives incorrect results on PyPy but not on CPython. So we don't want to hassle people on CPython until they have an alternative to move to, but on PyPy you probably want to know your code is broken even if there isn't a clear alternative.

xoviat replied:
I supposed it's fine as it is then.

xoviat commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -85,17 +85,56 @@ PyArray_Size(PyObject *op)
+    /* TODO: enable this once a solution for UPDATEIFCOPY
+     *  and nditer are resolved, also pending the fix for GH7054
+     */
+    if (DEPRECATE("PyArray_SetUpdateIfCopyBase is deprecated, use "
```
Also, this needs a version and date.

xoviat commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -387,26 +474,42 @@ array_dealloc(PyArrayObject *self)
+                }
+                else {
+                    PyErr_WriteUnraisable(Py_None);
+                }
```
You may need to `Py_DECREF` the `s` object, IIUC.

xoviat commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -2005,9 +2011,31 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
+             * Deprecate this (and remove the `else` from the following
+             * condition) once the flag itself is deprecated
+             */
+
```
This wording is a bit confusing. A more accurate statement would be "this logic branch is deprecated as of version X with date."

xoviat commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -208,18 +208,35 @@ arrayflags_dealloc(PyArrayFlagsObject *self)
+arrayflags_updateifcopy_get(PyArrayFlagsObject *self)
+{
+    PyObject *item;
+    if(DEPRECATE("UPDATEIFCOPY deprecated, use WRITEBACKIFCOPY instead") < 0) {
```
Version and date.

xoviat commented on numpy/core/src/multiarray/flagsobject.c:
```diff
@@ -291,6 +308,34 @@ arrayflags_updateifcopy_set(PyArrayFlagsObject *self, PyObject *obj)
                 "Cannot set flags on array scalars.");
         return -1;
     }
+    if(DEPRECATE("UPDATEIFCOPY deprecated, use WRITEBACKIFCOPY instead") < 0) {
```
Version and date.

njsmith commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -387,26 +474,42 @@ array_dealloc(PyArrayObject *self)
+        if (PyArray_FLAGS(self) & NPY_ARRAY_WRITEBACKIFCOPY)
+        {
+            char * msg = "WRITEBACKIFCOPY requires a call to "
+                "PyArray_ResolveWritebackIfCopy before array_dealloc is "
```
Should probably mention `DiscardWritebackIfCopy` here as another option.

ahaldane replied:
@mattip pinging this in case you missed it

mattip replied:
thanks

njsmith commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -2005,9 +2011,31 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
+             */
+
+            if (DEPRECATE("NPY_ARRAY_UPDATEIFCOPY (possibly via "
+                "NPY_ARRAY_INOUT_ARRAY or NPY_ARRAY_INOUT_FARRAY) is "
```
Probably simpler to phrase this as "NPY_ARRAY_UPDATEIFCOPY, NPY_ARRAY_INOUT_ARRAY, and NPY_ARRAY_INOUT_FARRAY are deprecated, use ..."

njsmith commented on doc/CAPI.rst.txt:
```diff
@@ -274,19 +274,25 @@ array, or part of some larger record array.  But, they may have other uses...
-    "misbehaved" array was not ``WRITEABLE`` to begin with then
-    ``PyArray_FromAny`` would have returned an error because ``UPDATEIFCOPY``
+    points to the "misbehaved" array (which is set read_only). If you use
+    this flag, you are must call ``PyArray_ResoveWritebackIfCopy`` before
```
"or `PyArray_DiscardWritebackIfCopy`"

njsmith commented on doc/release/1.14.0-notes.rst:
```diff
@@ -124,10 +124,45 @@ the result is always a view on the original masked array.
+visible change in behavior is that if compiled with ``-DDEPRECATE_UPDATEIFCOPY``
+or if run on PyPy, and writeback resolution has not occurred before calling
+``array_dealloc``, NumPy will now raise a DeprecationWarning.
+Change internal WRITEBACKIFCOPY (depreacates UPDATIFCOPY) semantics
```
"deprecates" misspelled

njsmith commented on numpy/core/include/numpy/ndarrayobject.h:
```diff
@@ -171,18 +171,27 @@ extern "C" CONFUSE_EMACS
+{
+    PyArray_DiscardWritebackIfCopy(arr);
+    Py_XDECREF(arr);
+}
```
I'm not sure how useful `NPY_NO_DEPRECATED_API` really is, but using it here isn't a terrible idea. We should probably call `DEPRECATE` too though.

njsmith reviewed (commented):
Left some minor comments, but overall this looks good. My one substantive question is: why is it that some of the new `DEPRECATE` calls are protected by `#ifdef DEPRECATE_UPDATEIFCOPY` and some are not? Maybe we talked about this before and I forgot (sorry!), but it seems like the obvious thing would be to use the `#ifdef` in all cases?

Once that and my and @xoviat's little text tweaks are fixed, I think this can go in.

mattip committed:
code review changes

xoviat commented:
Can you resolve the merge conflicts?

mattip commented:
Only one of the deprecate calls is protected by ``#ifdef DEPRECATE_UPDATEIFCOPY``, the one inside the deprecated function ``PyArray_SetUpdateIfCopyBase``, which unfortunately is still used by nditer (see issue #9714 for an in-depth discussion of what needs to be done to remove this). Once we have resolved that issue, we can remove the ``#ifdef DEPRECATE_UPDATEIFCOPY`` protection and truly deprecate ``PyArray_SetUpdateIfCopyBase``

xoviat commented:
That's what I thought.

mattip commented:
I can resolve merge conflicts, my github foo suggests that once I do so the review comments will be more difficult to resolve to lines of code, so please take a look first that I have done what you asked

mattip commented:
ahh, I see they get an "outdated" tag anyway so I will rebase and resolve

xoviat commented:
Just use the web editor rather than rebasing. This will probably be squashed anyway.

mattip committed:
clean up merge cruft

mattip committed:
Merge branch 'master' into refactor-updateifcopy2

xoviat commented:
@mattip Because of a Github bug, you need to merge twice with the web editor. Just repeat the same changes.

njsmith commented:
Oh, ok: so the plan is actually to immediately deprecate everything except the special case of nditer's internal usage? It's a bit confusing because `PyArray_SetUpdateIfCopyBase` is itself a public API with potentially lots of consumers, so if our goal is to deprecate everything except nditer, then the current `#ifdef` doesn't exactly do that... it's possible there are third-party packages whose only use of `UPDATEIFCOPY` is via `PyArray_SetUpdateIfCopyBase`. OTOH I don't know if any such packages actually exist. It [doesn't look like there are any packages in Debian like this](https://codesearch.debian.net/search?q=PyArray_SetUpdateIfCopyBase).

I guess one option would be to split `PyArray_SetUpdateIfCopyBase` into a public wrapper, which issues a deprecation warning, and a private function that actually does the work. Then we could drop the `#ifdef` from `PyArray_SetUpdateIfCopyBase`, and instead tweak nditer (for now) so that on the branch that currently calls `PyArray_SetUpdateIfCopyBase`, it instead does something like
```c
#ifdef PYPY
    PyErr_SetString(PyExc_RuntimeError, "...");
    return 0;
#endif
    internal_pyarray_setupdateifcopybase(...);
```

mattip commented:
is ``PyArray_SetUpdateIfCopyBase`` a public interface?

xoviat commented:
> is PyArray_SetUpdateIfCopyBase a public interface?

yes. 

> I guess one option would be to split PyArray_SetUpdateIfCopyBase into a public wrapper, which issues a deprecation warning, and a private function that actually does the work. 

IMHO I agree with this but as I noted, breaking the changes into separate PRs is better.

mattip committed:
Merge branch 'master' into refactor-updateifcopy2

njsmith commented:
@mattip: Unfortunately yes -- you can tell from the `NUMPY_API` comment just above it. (`NUMPY_API` is a magic string that's used to some preprocessor scripts to find API functions and export them.)

mattip commented:
``PyArray_SetUpdateIfCopyBase`` will be deprecated, it is replaced by ``PyArray_SetWritebackIfCopyBase`` ( I should have been able to answer the public question, both appear in ``numpy_api.py``) so once we can fix #9714 we can deprecate it entirely.

xoviat commented:
@njsmith Ready.

mattip commented:
@njsmith if I could get a thumbs-up on this someone else can do the merge. Can you at some point give a stronger approval than the "[I think this can go in](https://github.com/numpy/numpy/pull/9639#pullrequestreview-74934285)" above?

ahaldane commented on numpy/core/src/multiarray/calculation.c:
```diff
@@ -249,8 +250,9 @@ PyArray_ArgMin(PyArrayObject *op, int axis, PyArrayObject *out)
 
     Py_DECREF(ap);
-    /* Trigger the UPDATEIFCOPY if necessary */
+    /* Trigger the UPDATEIFCOPY/WRTIEBACKIFCOPY if necessary */
```
typo

mattip replied:
thanks

mattip committed:
fix from review

ahaldane commented:
OK so I just read through the problem. I'm not sure I caught all the subtleties, but at least I see the point and I see why this is an improvement.

So if this gets into 1.14, is the plan also to implement #9714 in time for 1.14?

And if I understand correctly, the discussion at this point is about exactly when to emit deprecation warnings. In 1.14 the old updateifcopy flag will still "work", just emit warnings. Is that right?

ahaldane commented:
Also, `PyArray_SetUpdateIfCopyBase` may be part of the API, but it doesn't seem to be documented on the api doc page.

mattip commented:
@ahaldane issue #9714 is a much smaller project so if there is consensus on the direction it could be ready for 1.14. The old flag and behaviour is still supported, and will emit warnings EXCEPT in the case of nditer (on CPython).

mattip committed:
fixes from review

mattip commented:
@ahaldane the fix for issue #9714 is a much smaller so if there is consensus on the direction it could be ready for 1.14. Please take a look and tell me if we should deprecate the use case or make nditer into a context manager. 

The old flag and behaviour is still supported, and will emit warnings, except for the cases mentioned in issue #9714 (for a more complete discussion see [this part of this pull request](https://github.com/numpy/numpy/pull/9639#discussion_r149516463)

mattip commented:
@ahaldane the fix for issue #9714 is a much smaller so if there is consensus on the direction it could be ready for 1.14. Please take a look and tell me if we should deprecate the use case or make nditer into a context manager. 

The old flag and behaviour is still supported, and will emit warnings, except for the cases mentioned in issue #9714 (for a more complete discussion see [this part of this pull request](https://github.com/numpy/numpy/pull/9639#discussion_r149516463)

mattip commented:
@ahaldane the fix for issue #9714 is a much smaller so if there is consensus on the direction it could be ready for 1.14. Please take a look and tell me if we should deprecate the use case or make nditer into a context manager. 

The old flag and behaviour is still supported, and will emit warnings, except for the cases mentioned in issue #9714 (for a more complete discussion see [this part of this pull request](https://github.com/numpy/numpy/pull/9639#discussion_r149516463)

ahaldane commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -895,10 +896,14 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
 #define NPY_ARRAY_INOUT_FARRAY (NPY_ARRAY_FARRAY | \
                                 NPY_ARRAY_UPDATEIFCOPY)
+#define NPY_ARRAY_INOUT_FARRAY2 (NPY_ARRAY_FARRAY | \
+                                NPY_ARRAY_WRITEBACKIFCOPY)
```
So presumably this `2` version is only needed during the deprecation phase, afterwards we will redefine `NPY_ARRAY_INOUT_FARRAY` to use `NPY_ARRAY_WRITEBACKIFCOPY`.

njsmith replied:
That's one option, but it's a bit risky since we don't have any way to know when all the old code using `NPY_ARRAY_INOUT_FARRAY` is gone. We might just keep the `2` version forever – it's not pretty but it's safe and works.

ahaldane replied:
yeah we can leave the 2 version forever, but after the deprecation phase no one needs to use it any more.

ahaldane commented:
What do you think of changing the release note as follows:
```
``UPDATEIFCOPY`` semantics deprecated in favor of ``WRITEBACKIFCOPY``
--------------------------------------------------------------------
The ``UPDATEIFCOPY`` flag is deprecated, and replaced by a new flag
``WRITEBACKIFCOPY``. In the C-api, the ``NPY_ARRAY_UPDATEIFCOPY`` flag is
deprecated, and should be replaced with ``NPY_ARRAY_WRITEBACKIFCOPY`` and
semantics.

Code using these flags must be updated to account for changed semantics, as
follows:. When ndarrays are created with either the deprecated ``UPDATEIFCOPY``
or new ``WRITEBACKIFCOPY`` flags, a temporary copy of the data is created which
is eventually written back to the original array. For ``UPDATEIFCOPY``, the
writeback to the original array occurs at ndarray deallocation. For
``WRITEBACKIFCOPY``, a new ``PyArray_ResolveWritebackIfCopy`` function must be
explicitly called before deallocation (or, if an error occurred,
``PyArray_DiscardWritebackIfCopy``). Numpy now mostly uses the
``WRITEBACKIFCOPY`` mechanism internally, but still uses ``UPDATEIFCOPY`` in
nditer use for back-compatibility of the python nditer interface.

In python code, if numpy is compiled with ``-DDEPRECATE_UPDATEIFCOPY`` or if
run on PyPy, ``DeprecationWarning`` warnings will be issued on use of
``UPDATEIFCOPY`` (eg, when nditer is used) if writeback resolution has not
occurred before calling ``array_dealloc``. In the future this warning will
always be issued, once nditer stops using ``UPDATEIFCOPY``.

In C code, calling ``PyArray_SetUpdateIfCopyBase`` will now always issue a
``DeprecationWarning`` and should be replaced by
``PyArray_SetWritebackIfCopyBase``. Similarly, calls to ``PyArray_XDECREF_ERR``
should be replaced by ``PyArray_DiscardWritebackIfCopy``.  Calling ndarray
creation functions such as ``PyArray_FromArray`` where flags uses
``NPY_ARRAY_UPDATEIFCOPY`` will also raise a ``DeprecationWarning``. 

Note that during the deprecation period, the ``NPY_ARRAY_INOUT_ARRAY`` and
``NPY_ARRAY_INOUT_FARRAY`` flags should be replaced by
``NPY_ARRAY_INOUT_ARRAY2`` and ``NPY_ARRAY_INOUT_FARRAY2``, which behave
similarly but use the new `WRITEBACKIFCOPY` semantics.
```

I want to make it easy for downstream projects to know what to change.

mattip committed:
extend the release note

njsmith commented:
@ahaldane: Are you happy?

xoviat commented:
> And if I understand correctly, the discussion at this point is about exactly when to emit deprecation warnings. In 1.14 the old updateifcopy flag will still "work", just emit warnings. Is that right?

That's correct. Generally the deprecation procedure is to immediately emit warnings as soon as the deprecations occur. We can't do that here because there is still some internal code that calls this, but that doesn't work on pypy so we do that anyway. IMHO the next PR should rename the function so that warnings can also be emitted on CPython.

ahaldane commented:
Yeah looks good to me.

njsmith commented:
Okay, let's do this...

eric-wieser commented on doc/source/reference/c-api.array.rst:
```diff
@@ -3432,13 +3468,25 @@ Miscellaneous Macros
+    :c:data:`NPY_ARRAY_UPDATEIFCOPY` flag. Also resets the
+    :c:data:`NPY_ARRAY_WRITEABLE` flag on the base object. This is
+    useful for recovering from an error condition when
+    writeback semantics are used, but will lead to wrong results.
```
Wording isn't great here - I read this as "This is useful ..., but will lead to wrong results", which is just confusing

mattip replied:
how about "useful for recovering from an exception when writeback semantics are used. Pending changes to the base array will be thrown away"

eric-wieser commented on numpy/core/code_generators/cversions.txt:
```diff
@@ -36,5 +36,6 @@
 # Version 11 (NumPy 1.13) Added PyArray_MapIterArrayCopyIfOverlap
-# Version 11 (NumPy 1.14) No Change
+# Version 11 (NumPy 1.14) Added PyArray_ResolveWritebackIfCopy,
+# PyArray_SetWritebackIfCopyBase and deprecate PyArray_SetUpdateIfCopyBase.
```
Shouldn't this get a new hash on the last line? Or does that only happen at release time? @charris?

At any rate, I think the comment should go below the last hash in the file.

charris replied:
It gets checked and updated at release time, but should be updated on the fly. IIRC, that was how it was done with the last addition in 1.13.

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -385,26 +475,51 @@ array_dealloc(PyArrayObject *self)
+                s = PyUnicode_FromString(msg);
+#endif
+                if (s) {
+                    PyErr_WriteUnraisable(s);
```
This is not correct usage - the argument should be the _context_ - the message is already printed along with the exception type. Right now, the message is printed twice.

Off the top of my head, `(PyObject *)&PyArray_Type` would be a safe context to pass. In theory you could pass `self`, but calling `array_repr` sounds a little too risky.

if nothing else, passing `NULL` would cause the message alone to be printed.

eric-wieser replied:
If we used the py3 `tp_finalize` slot, then we could pass `ndarray.__del__` as the context.

mattip replied:
> Right now, the message is printed twice

This code should only run if ``DEPRECATE`` (which is [``PyErr_WarnEx``](https://docs.python.org/3/c-api/exceptions.html#issuing-warnings) ) raises an exeption, in which case the exception will be ignored since this is a deallocator. In order to be sure the error is emitted, I wanted to print it to ``stderr``, and the documentation seems to indicate that is what [``PyErr_WriteUnraisable``](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable) does. Should I print to stderr and then emit ``PyErr_WriteUnraisable`` as well, or simply skip it altogether? Since there was no example in NumPy I looked at how ``PyErr_WriteUnraisable`` is used in Pandas as a hint.

eric-wieser replied:
> the documentation seems to indicate that is what PyErr_WriteUnraisable does

Correct, but it does this by looking at `sys,exc_info`, not by looking at its argument

mattip replied:
I think I understand, the entire ``if()`` clause should be replaced by the one-liner 
```
PyErr_WriteUnraisable((PyObject *)&PyArray_Type);
```
How should I proceed since this pull request has been merged? Issue another pull request based off HEAD?

mattip replied:
hmm, it seems cpython prints both the context from ``PyErr_Fetch`` and ``repr(obj)`` for a call to ``PyErr_WriteUnrasiable`` but PyPy only prints the ``repr(obj)`` without the context. I will file an issue there

eric-wieser replied:
I'd personally go for a new PR based on the last commit of this PR, but going off head would work fine too.

eric-wieser commented:
> the documentation seems to indicate that is what PyErr_WriteUnraisable does

Correct, but it does this by looking at `sys,exc_info`, not by looking at its argument
